Gitcoin bounty

d80ep08th opened the pull request:
This PR should close Issue #1

d80ep08th committed:
UI made with xamarin form shapes api

d80ep08th committed:
Four arcs updated as named objects

I did this using x:Name = "" in the tags I wanted to manipulate

d80ep08th committed:
Amount triggers Arc Length

Changing the amount value will dynamically change lengths of the 4 arcs according to the percentages in render(amount)

d80ep08th committed:
Update MainPage.xaml.cs

knocte commented:
@d80ep08th good work so far! This is not finished yet, though, because I've found a little problem with it, but bear with me, I'm sure you will be able to resolve it very easily. I'll comment on the diff.

knocte commented on ShapesBalanceXamFormsApp/MainPage.xaml.cs:
```diff
@@ -5,14 +5,202 @@
+
+        }
+
+        public void makePies(double P0, double P1, double P2, double P3)
```
@d80ep08th why is this hardcoded to 4 elements? Take in account that the screenshot in https://github.com/nblockchain/ShapesBalanceXamFormsApp/issues/1 had 7 colors, and the comment in which we mention the "parameterized" requirement states 3 amounts.

Now, I'm not trying to ask you to change it from 4 parameters to 3 or to 7. What I'm trying to tell you is that this function should receive a single collection parameter, so that it can render pies of any number of arcs. In particular, the signature should have been something like:

```
void makePies(IEnumerable<double> amounts)
```

Only issue with the above would be the colours; and you told me you couldn't choose them dynamically. Well, I disagree, I think you can; however, that is a minor problem that I'm willing to help you with later. Let's fix this issue first with the amounts, and then I will jump myself to help you with the colours, sounds good?

d80ep08th replied:
Yeah sounds good. 

Well,  I figured out the issue with colors, I was using "Color.Black"  to render Black  which throwed an error when I tried to build and the solution was to use  Brush instead of Color i.e.  "Brush.Black".

Now as Im trying to make changes on my side considering what I understood from your review, here's what I am doing:

1. I hardcoded 10 arc segments
2. dynamically allocated their colors using Brush 
3.  built a function to render the arcs which reduced the code.
4. Finally , based on the amount  I am trying to vary the number of arcs between 10 and 6 . 

Example : 
0 < amount < 1000 , render 6 arcs
1000<amount<2000, render 7 arcs
2000<amount<3000, render 8 arcs
3000<amount<4000, render 9 arcs
4000<amount<5000, render 10 arcs

Is this according to what you are thinking?

knocte replied:
> Is this according to what you are thinking?

TBH, not really. I asked you to not hardcode and you're already thinking of hardcoding it for an arbitrary range of 6 to 10 arcs, why? It should work with any number of arcs. Just think about it, really, why would a wallet assume that a crypto user has as many as 6 currencies? She could have 2, or 1, or 3 (the most likely, probably not more than that), or whatever the number! But definitely not just 6-10. I don't see why you need to hardcode any arc. And I don't see why you need to place `<Path>` elements in the XAML, just generate them with C#, no?

knocte replied:
@d80ep08th Also, I don't really understand what do you mean with those 1000,2000,3000,.. amounts. We're talking about percentages here. For example, let's say a user has 200 of x(e.g. red), 100 of y (e.g. blue), and 100 of z (e.g. yellow), then you have to convert this to a chart with 3 arcs, in which one is a semi-circle (200 out of 400 -> 50%) and the other two are 25% each.

d80ep08th replied:
I think I understand now:

So if the wallet contains : 200 of X(Red) , 100 of Y (Blue ), 100 of Z (Yellow)

The Account balance is basically, the sum of the amount of X , Y , Z after conversion to Euros

And the colors around the Account balance basically represent the different cryptos in which the  total account balance is divided.

knocte replied:
>  the sum of the amount of X , Y , Z after conversion to Euros

Yes, 200 and 100 are all euro amounts. We never referred to crypto amounts anywhere (because if we did, you would need the exchange rate to euros as an input too).

> And the colors around the Account balance basically represent the different cryptos in which the total account balance is divided.

Yes

d80ep08th replied:
I just made changes that resolves this issue. I made everything Dynamic

>  We never referred to crypto amounts anywhere (because if we did, you would need the exchange rate to euros as an 
> input too).


I think that is where I got confused, it did not make sense to me why would I choose random fractions of the total amount to make the pie

knocte replied:
Ok great!, the approach looks good now.

However, I have found a bug. I changed the amounts to this:

```
            yield return new Wallet(400, Brush.Orange);
            yield return new Wallet(40, Brush.Yellow);
            yield return new Wallet(20, Brush.Green);
            yield return new Wallet(20, Brush.Blue);
```

And this was the result:

<img width="468" alt="Screenshot 2022-05-18 at 12 55 55 PM" src="https://user-images.githubusercontent.com/331303/168960135-596800d8-54fa-4c8d-b276-8b33f21aa233.png">

d80ep08th replied:
Oh snap , looks like I need to normalize them values before they hit the func

d80ep08th replied:
I just committed changes that should rectify the circle's geometry from being messed up

d80ep08th replied:
I used these values in the code I committed
```
            yield return new Wallet(400, Brush.Green);
            yield return new Wallet(40, Brush.Blue);
            yield return new Wallet(20, Brush.Gold);
            yield return new Wallet(20, Brush.Olive);
            yield return new Wallet(20, Brush.Black);
```

knocte replied:
@d80ep08th seems the circle is now always rendered correctly, however the arc sizes are not right.

See this example:

```
            yield return new Wallet(50, Brush.Orange);
            yield return new Wallet(5, Brush.Yellow);
            yield return new Wallet(5, Brush.Blue);
```

In the case above, the arc for orange should be much bigger than the other two; however, this is what I see:

<img width="418" alt="Screenshot 2022-05-18 at 5 57 36 PM" src="https://user-images.githubusercontent.com/331303/169013004-c6b9255a-d85a-4a37-b260-710f9ad730e9.png">

d80ep08th replied:
I realized that the larger values will generate angles greater than 180 degrees which should trigger the isLargeArc boolean to turn true, So instead of using a function for normalizing the larger values I normalized the smaller values before it hits the function by introducing a variable `normalize`

d80ep08th replied:
Just made a commit that should generate perfectly sized arcs wrt the values taking into consideration all test-cases.

knocte replied:
@d80ep08th thanks for your persistence on this matter.

Unfortunately, this is still not correct. I just tested the same testcase that I posted in my last update:

```
            yield return new Wallet(50, Brush.Orange);
            yield return new Wallet(5, Brush.Yellow);
            yield return new Wallet(5, Brush.Blue);
```

As you can see from the above, the total amount is 60. Orange is 50 so it is a percentage of (100*50/60=)**83.3%**, but the chart rendered doesn't reflect this (as you can see, the arc for orange is less than 50% of the assets):

<img width="465" alt="Screenshot 2022-05-19 at 2 14 30 PM" src="https://user-images.githubusercontent.com/331303/169223203-f03828b0-0dac-46c8-a52d-a73d1e8e2426.png">

I don't know what you're trying to with your normalization code but you need to go back to basics. Convert amounts to percentages first, and then just draw the arcs according to their size.

d80ep08th replied:
Hey @knocte  , please check if the changes I committed solves the issue

knocte replied:
Seems it's working properly now!

d80ep08th replied:
Please merge it then

knocte replied:
Yes I will. I paid out the bounty first, do you see it?

d80ep08th replied:
Yes I see it, thank you!

d80ep08th committed:
Removing all Hardcoded Shape Elements

d80ep08th committed:
Clearing Extra Space

d80ep08th committed:
Dynamically Building  Shapes

d80ep08th committed:
Normalized amount values before rendering pie

d80ep08th committed:
Alternative approach to normalize the pie

d80ep08th committed:
New function that normalizes arcs

The new normalize function ranks the values first , saves the ranks in a list at the same positions of the crypto amount value and then creates pies from the ranks list. 

Also. new variable `gap` to manipulate gap between arcs

d80ep08th committed:
Update MainPage.xaml.cs

d80ep08th committed:
removed normalize function

d80ep08th committed:
Full circle

Makes a circle when a single value is passed

d80ep08th committed:
Smallest gap

Smallest gap that can be maintained between two arcs

d80ep08th committed:
Updated condition to render small arcs

d80ep08th committed:
Updated conditions that renders the arc

e.g. 99&1 being converted to 98&2, but 99.1&0.9 being converted to 100&0

d80ep08th committed:
Revert "Updated conditions that renders the arc"

This reverts commit 5ddf4663b7d1b0d374d7221626cff97da91e0d95.

d80ep08th committed:
Renamed variable to a better name

d80ep08th committed:
Refactoring : removing unneeded return Arc

d80ep08th committed:
Refactoring: Moved code to the method setArcAngle

d80ep08th committed:
Added unit test
